[MRG] Block-wise silhouette calculation to avoid memory consumption

jnothman opened the pull request:
Superseding #1976 which breaks the problem down by cluster, here we simply break the input down by fixed-size blocks of rows. Reports of memory issues when calculating silhouette include #7175 and [mailing list](http://sourceforge.net/mailarchive/message.php?msg_id=30828533). Also incorporates #6089's fix for #5988; and adds a test for `silhouette_samples` self-plagiarised from #4087.

potential enhancements:
- [x] work out what to do to avoid using `np.ufunc.at` in numpy<1.8
- [x] the minor TODOs mentioned in the code
- [x] benchmark to ensure this is not much slower than incumbent implementation. [See  #below](https://github.com/scikit-learn/scikit-learn/pull/7177#issuecomment-239166677).
- [x] a test for `silhouette_samples` (none exists!); could perhaps be a separate PR...
- [x] support parallelised computation over blocks?; could be a separate PR
- [x] allow users to specify `block_size` by target memory usage, e.g. `block_size='1G'`. Probably out of scope of this PR.
- [x] use `block_size=100` or `block_size='1G'` or some other sensible constant as the default to avoid memory errors with default parameters, and to improve default speed.
- [x] add more inline comments
- [ ] add to what's new
- [ ] squash carefully to ensure attribution of #6089

Sentient07 committed:
Reverted the change, added regression test

reverted the comment

Resolved merge conflicts

jnothman committed:
ENH block_size for memory efficiency in silhouette

jnothman committed:
Merge remote-tracking branch 'upstream/pr/6089' into silhouette-chunks

jnothman committed:
DOC add versionadded to new parameter

jnothman committed:
FIX use bincount instead of np.add.at for old numpy

jnothman committed:
ENH use unary pairwise_distances where possible

jnothman commented:
Benchmarks with `block_size=None`, attempting to follow #6151:

```python
import numpy as np
from sklearn.metrics.cluster import silhouette_score
n_features = 100
for n_samples in [10000, 1000, 100]:
	for n_labels in [4, 6, 100]:
		print('n_features', n_features, 'n_samples', n_samples, 'n_labels', n_labels)
		X = np.random.rand(n_samples, n_features)
		y = np.random.randint(n_labels, size=n_samples)
		%timeit silhouette_score(X, y)
```

```
master                                        block_size=None                               block_size=100
n_features 100 n_samples 10000 n_labels 4     n_features 100 n_samples 10000 n_labels 4     n_features 100 n_samples 10000 n_labels 4
1 loop, best of 3: 3.28 s per loop            1 loop, best of 3: 4.03 s per loop            1 loop, best of 3: 1.61 s per loop
n_features 100 n_samples 10000 n_labels 6     n_features 100 n_samples 10000 n_labels 6     n_features 100 n_samples 10000 n_labels 6
1 loop, best of 3: 3.16 s per loop            1 loop, best of 3: 4.12 s per loop            1 loop, best of 3: 1.64 s per loop
n_features 100 n_samples 10000 n_labels 100   n_features 100 n_samples 10000 n_labels 100   n_features 100 n_samples 10000 n_labels 100
1 loop, best of 3: 3.48 s per loop            1 loop, best of 3: 4.05 s per loop            1 loop, best of 3: 1.61 s per loop
n_features 100 n_samples 1000 n_labels 4      n_features 100 n_samples 1000 n_labels 4      n_features 100 n_samples 1000 n_labels 4
100 loops, best of 3: 16.9 ms per loop        10 loops, best of 3: 30.2 ms per loop         100 loops, best of 3: 15.8 ms per loop
n_features 100 n_samples 1000 n_labels 6      n_features 100 n_samples 1000 n_labels 6      n_features 100 n_samples 1000 n_labels 6
100 loops, best of 3: 18.3 ms per loop        10 loops, best of 3: 30 ms per loop           100 loops, best of 3: 15.7 ms per loop
n_features 100 n_samples 1000 n_labels 100    n_features 100 n_samples 1000 n_labels 100    n_features 100 n_samples 1000 n_labels 100
1 loop, best of 3: 226 ms per loop            10 loops, best of 3: 30.8 ms per loop         100 loops, best of 3: 16.7 ms per loop
n_features 100 n_samples 100 n_labels 4       n_features 100 n_samples 100 n_labels 4       n_features 100 n_samples 100 n_labels 4
1000 loops, best of 3: 688 µs per loop        1000 loops, best of 3: 570 µs per loop        1000 loops, best of 3: 608 µs per loop
n_features 100 n_samples 100 n_labels 6       n_features 100 n_samples 100 n_labels 6       n_features 100 n_samples 100 n_labels 6
1000 loops, best of 3: 1.01 ms per loop       1000 loops, best of 3: 572 µs per loop        1000 loops, best of 3: 600 µs per loop
n_features 100 n_samples 100 n_labels 100     n_features 100 n_samples 100 n_labels 100     n_features 100 n_samples 100 n_labels 100
10 loops, best of 3: 57.9 ms per loop         1000 loops, best of 3: 591 µs per loop        1000 loops, best of 3: 587 µs per loop
```

defaulting to `block_size=100` is looking pretty good too...

jnothman committed:
DOC explicit block_size parameter in silhouette_score

jnothman committed:
TST test silhouette_samples explicitly

jnothman committed:
ENH support n_jobs in silhouette_score

and add comments

jnothman committed:
DOC update block_size description given n_jobs

raghavrv commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -56,6 +60,18 @@ def silhouette_score(X, labels, metric='euclidean', sample_size=None,
 
+    block_size : int, optional
+        The number of rows to process at a time to limit memory usage to
+        O(block_size * n_samples). Default is n_samples.
```
nitpick: Double backticks

raghavrv commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -144,6 +189,18 @@ def silhouette_samples(X, labels, metric='euclidean', **kwds):
 
+    block_size : int, optional
+        The number of rows to process at a time to limit memory usage to
+        O(block_size * n_samples). Default is n_samples / n_jobs.
```
nitpick: codeformatting

jnothman committed:
DOC docstring formatting

jnothman commented:
Instead of allowing the user to specify size as a unit of memory usage, I am wondering if `block_size` should specify maximum number of cells created by the `pairwise_distances` operation, rather than number of rows. I.e. the latter is the former divided by `n_samples`.

jnothman commented:
I've now tried to just have the user specify the pairwise distance memory consumption in bytes...

jnothman committed:
ENH specify silhouette block size in bytes

raghavrv commented:
> I've now tried to just have the user specify the pairwise distance memory consumption in bytes...

If it needs to be in memory units, could we have it in MBs?

jnothman commented:
> If it needs to be in memory units, could we have it in MBs as users will need to enter a smaller number?

It doesn't need to be, but I think it is a more practical data-invariant measure. Yes, can have in MiB

jnothman committed:
block_size specified in MiB

jnothman commented:
@raghavrv now in MiB

tguillemot commented:
>Instead of allowing the user to specify size as a unit of memory usage, I am wondering if block_size should specify maximum number of cells created by the pairwise_distances operation, rather than number of rows. I.e. the latter is the former divided by n_samples.

@jnothman I'm a little bit late but +1 to have the memory usage. I think it's an easier measure to specify. And also +1 for MiB.

jnothman commented:
You're earlier than everyone but @raghavrv. Thanks for the input. That's where it's converged anyway.

tguillemot commented:
@jnothman I do a review in a few moment.

tguillemot commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -103,10 +124,39 @@ def silhouette_score(X, labels, metric='euclidean', sample_size=None,
+
+
+def _process_block(X, labels, start, block_n_rows, block_range, add_at,
+                   label_freqs, metric, kwds):
```
Just a niptick : _process_block -> _silhouette_process_block.
I know it's a private function but maybe it will be easier for comprehension if you put a little docstring of the arguments ???

jnothman replied:
Fair enough. Will do

raghavrv commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -56,6 +65,18 @@ def silhouette_score(X, labels, metric='euclidean', sample_size=None,
         <sklearn.metrics.pairwise.pairwise_distances>`. If X is the distance
         array itself, use ``metric="precomputed"``.
 
+    block_size : int, optional, default=64
```
Could it be `memory_per_job` maybe?

tguillemot replied:
In the GPU computing language (OpenCl, Cuda),`block_size` makes sense.

jnothman replied:
I'm not committed to it. (Nor am I, by the way, committed to this being per job. I'm happy for it to be divided by the jobs.) But it's not explicitly the maximum memory consumption either. `add_at` contributes `n_samples * n_clusters` further ints.

raghavrv replied:
Okay then can we have it as `memory_usage` and let it denote the total memory consumed for all the jobs. And add a note in the docstring stating that this is not the maximum memory as there could be some more overheads?

Because as a user, I'd blindly set this to my ram maximum. And `n_jobs` to number of processors I have...

jnothman replied:
If you're an average user, I'm better off not making it a parameter and hard-coding something sensible! :P

raghavrv replied:
lol

tguillemot replied:
Indeed :)

tguillemot commented:
The performances are similar on my computer.

raghavrv commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -56,16 +77,69 @@ def test_silhouette():
+    y = [1, 1, 2]
+    assert_raise_message(ValueError, 'block_size should be at least n_samples '
+                         '* 8 bytes = 1 MiB, got 0',
+                         silhouette_score, X, y, block_size=0)
```
Should we instead have the `block_size` as 0 and let it denote the master setup of use all the memory?

raghavrv replied:
Or -1 like `n_jobs`?

jnothman replied:
For what benefit? The default 64MB will run a 2896 sample problem or smaller in a single block. For problems much larger than that, you're likely to benefit from splitting the problem up as suggested by our benchmark which shows 2x speedup from "use all memory" for a dataset less than 4x that size (and >9x the number of pairwise calculations). Yes, this is only my machine, but it's hard to imagine why we should suggest using all memory possible to the user.

jnothman replied:
Not dissimilar to `n_jobs=-2` often being a better choice than `n_jobs=-1`

tguillemot replied:
Why do not select automatically `block_size` as the `min(64MB, np.ceil(n_samples * BYTES_PER_FLOAT * 2 ** -20)` and don't let the user choose a specific value ?

jnothman replied:
That would be identical to just setting it to 64, no? That's tempting, especially because this isn't a learning routine. I don't expect my benchmarks to be optimal, certainly not for any particular platform, or with `n_jobs != 1` (using the default or some other parallel backend); hence my inclination to keep it public. I'll have to chew on this.

jnothman committed:
document parameters to silhouette helper

jnothman commented:
I'd appreciate another opinion on this: should we make a memory efficiency parameter constant, given that we know the constant is sensible on some modern platforms (and hard to explain to the user), or directly configurable by the user? In the context of a learning algorithm, configurability seems more valuable, but for an evaluation metric, is it sensible to just say: 64 MiB is the max pairwise distance evaluation available. Also, should this be `64 // n_jobs`? (perhaps I need to benchmark the interaction with `n_jobs`.)

Not sure who to ask: @gaelvaroquaux, @mechcoder, @amueller?

jnothman committed:
FIX pass n_jobs from silhouette_score

Also use threading for parallelism

jnothman commented:
Fixed a problem where `n_jobs` wasn't being used, and changed to use threading backend for parallelism (only marginal gains, but I've not seen multiprocessing be worth the overhead here).

raghavrv commented:
Sorry for piggy backing on this conversation... How do you usually definitively decide between multiprocessing and threading? As joblib memmaps the data I don't see why multiprocessing is any worse than threading...

jnothman commented:
I don't have any good answers for you, @raghavrv, but for the scale of problem and processor I've tried, multiprocessing always results in slower times. Afaik, threading is a good option when there's a lot of work in non-GIL operations, and when copying the inputs and outputs competes with the calculation for cost. I'm considering just dropping `n_jobs` support as ineffectual until someone can show me it actually benchmarks a more-than-marginal benefit.

amueller commented:
@jnothman I think for blocked computations, we set "reasonable" defaults and allow configuration, IIRC. Do you want a review on this? It doesn't seem high priority to me. [OT: why are you using this score?]

jnothman commented:
I'd like review on this, but no I'm not using silhouette personally. I worked on it because an issue came in, and because I wanted to ascertain that #1976 could be improved upon. I also think we should be reviewing the potential blockwise computations everywhere that `pairwise_distances` is used, such as KNN. Perhaps I should make an issue for that.

jnothman commented:
To be sure, it's not high priority, except that it closes a few issues and avoids more appearing over the next release.

jnothman commented:
Also, #5988 should be fixed.

raghavrv commented:
Could you refactor #7438 out of this PR and make a rebase. I can try and take a deeper look at this before the weekend :)

raghavrv commented:
@jnothman Ping :)
